app/components/channel_loader/channel_loader.js

```diff
@@ -9,15 +9,79 @@ import {
 } from 'react-native';
 import Placeholder from 'rn-placeholder';
 
+import EventEmitter from 'mattermost-redux/utils/event_emitter';
+
+import CustomPropTypes from 'app/constants/custom_prop_types';
 import {changeOpacity, makeStyleSheetFromTheme} from 'app/utils/theme';
 
 export default class ChannelLoader extends PureComponent {
     static propTypes = {
+        actions: PropTypes.shape({
+            handleSelectChannel: PropTypes.func.isRequired,
+            markChannelAsViewed: PropTypes.func.isRequired,
+            markChannelAsRead: PropTypes.func.isRequired,
+            setChannelLoading: PropTypes.func.isRequired,
+        }).isRequired,
         backgroundColor: PropTypes.string,
         channelIsLoading: PropTypes.bool.isRequired,
+        maxRows: PropTypes.number,
+        style: CustomPropTypes.Style,
         theme: PropTypes.object.isRequired,
     };
 
+    static defaultProps = {
+        maxRows: 6,
+    };
+
+    state = {
+        switching: false,
+    };
+
+    static getDerivedStateFromProps(nextProps, prevState) {
+        if (!nextProps.channelIsLoading && prevState.switch) {
+            return {
+                switch: false,
+                channel: null,
+                currentChannelId: null,
+            };
+        }
+
+        return prevState;
+    }
+
+    componentDidMount() {
+        EventEmitter.on('switch_channel', this.handleChannelSwitch);
+    }
+
+    componentWillUnmount() {
+        EventEmitter.off('switch_channel', this.handleChannelSwitch);
+    }
+
+    componentDidUpdate() {
+        if (this.state.switch) {
+            const {
+                handleSelectChannel,
+                markChannelAsRead,
+                markChannelAsViewed,
+                setChannelLoading,
+            } = this.props.actions;
+
+            const {channel, currentChannelId} = this.state;
+
+            setTimeout(() => {
+                handleSelectChannel(channel.id);
+
+                // mark the channel as viewed after all the frame has flushed
+                markChannelAsRead(channel.id, currentChannelId);
+                if (channel.id !== currentChannelId) {
+                    markChannelAsViewed(currentChannelId);
+                }
+
+                setChannelLoading(false);
+            }, 250);
```

Comment:
So we are adding a hardcoded delay and UX slowdown just to show the ChannelLoader? not a big fan of this solution

Reply:
This is odd. i can live with loader the first time but a recurring channel switch does not require a loader as we do have the information.  I think we should re-consider later about loader for loaded channels.

Reply:
that's true when the channel has been previously loaded and the posts haven't been cleaned up, we can look at improving this further in any case in a later PR, feel free to investigate and submit one ;)